packages/intrepid2/src/Discretization/Basis/Intrepid2_DerivedBasis_HGRAD_TRI.hpp

```diff
@@ -0,0 +1,398 @@
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Kyungjoo Kim  (kyukim@sandia.gov),
+//                    Mauro Perego  (mperego@sandia.gov), or
+//                    Nate Roberts  (nvrober@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+
+/** \file   Intrepid2_DerivedBasis_HGRAD_TRI.hpp
+    \brief  Implementation of H(grad) basis on the triangle that is templated on H(grad) on the line.
+    \author Created by N.V. Roberts.
+ 
+ H(grad) on the triangle, defined following the construction of Fuentes et al.
+ 
+ This class requires the template argument to be a hierarchical basis on the line; the construction here
+ will not be nodal even if the line basis provided is nodal.  (It would be nodal on the edges, but even
+ the edge functions would not be nodal on the interior of the triangle.)  Therefore, an exception is thrown
+ at construction if a non-hierarchical basis is provided.
+ 
+ In contrast to the derived bases defined on the quadrilateral and hexahedron, the interior (face) functions
+ defined here are defined independently of the line basis provided as a template argument.  Following Fuentes
+ et al., the interior functions are defined as Jacobi polynomials.
+ 
+ Eventually, we hope to develop a derived basis where the interior functions are defined in terms of the
+ functions defined on the line.  Ideally, we could define a generic "simplicial extension" that could be
+ used to define bases on the triangle, tetrahedron, and pyramid, as well as higher-dimensional simplices.
+ Ideally, this would also preserve hierarchical/nodal status.  As far as we know, however, such an extension
+ has not yet been constructed in the literature; for now, it remains an aspiration for us.
+ 
+ */
+
+#ifndef Intrepid2_DerivedBasis_HGRAD_TRI_h
+#define Intrepid2_DerivedBasis_HGRAD_TRI_h
+
+namespace Intrepid2
+{
+  // TODO: place SimplexTopologyMap in its own file
+  class SimplexTopologyMap
+  {
+    shards::CellTopology baseTopo_;
+  public:
+    SimplexTopologyMap(const shards::CellTopology &baseTopo)
+    :
+    baseTopo_(baseTopo)
+    {}
+    
+    /** \brief  Map from component subcell ordinals to the corresponding composite subcell ordinal.
```

Comment:
Not sure I understand the purpose of this function. Can you make an example?

Reply:
@mperego, this whole file is cruft from my earlier effort.  I was hoping to come up with a generic mechanism by which one could generate a triangle basis from a basis defined on the edge.  I've not *totally* given up hope on this idea, but you'd have to specify some additional constraints on the triangle basis (e.g. symmetry under rotations).  In any case, this is not what ESEAS does; they make some convenient but arbitrary choices for "blending" functions (they use Jacobi polynomials).  I ended up imitating their implementation pretty faithfully, which has the advantage that I was able to do some thorough cross-code verification.

Sorry for leaving the crufty file in there; I'll delete it.

Reply:
Oh OK, I see. No worries!